# Zero to Hero

## 1.06 Hackeando Kaggle

Voy a **hackear** Kaggle  para descubrir cuantos  "BAJA+2"  hay en el mes 202101, que es donde  *no* tengo la clase

La idea es subir a Kaggle una prediccion que tenga TODOS 1,  que se le envíe estímulo a todos
<br>
Es obvio que esa prediccion va a dar una pésima ganancia
<br>
Pero me va a permitir tener un sistema de dos ecuaciones con dos incognitas, y despejando tendre la cantidad de BAJA+2 que hay en los datos del futuro

In [1]:
rm( list=ls())
gc()

library( "data.table")   #cargo la libreria  data.table
library( "rpart")  #cargo la libreria  rpart
library( "rpart.plot")

options(repr.plot.width=20, repr.plot.height=10) 
setwd(".\\")  #Aqui se debe poner la ruta de la PC local

used    (Mb) gc trigger (Mb) max used (Mb)
Ncells  614700 32.9 1321401    70.6 1321401  70.6
Vcells 1093594  8.4 8388608    64.0 1632451  12.5

Warning message:
"package 'data.table' was built under R version 4.2.3"
Warning message:
"package 'rpart' was built under R version 4.2.3"
Warning message:
"package 'rpart.plot' was built under R version 4.2.3"


Leo el dataset

In [2]:
dataset <- fread("./datasets/dataset_pequeno.csv")   #cargo el dataset

Cuento la clase

In [3]:
dataset[  , .N, list( foto_mes, clase_ternaria) ]

foto_mes clase_ternaria N     
1 202107   CONTINUA       162336
2 202107   BAJA+1           1078
3 202107   BAJA+2           1268
4 202109                  165237

Defino  el data.table  dfuturo que va a tener solo los datos de  202109

In [4]:
dfuturo <-  dataset[ foto_mes==202109 ]

cuento la cantidad de lineas del dataset

In [5]:
nrow( dfuturo )

[1] 165237

Creo un vector que sean los  ids

In [6]:
vector_ids  <-   dfuturo[ , numero_de_cliente]

In [7]:
head( vector_ids)

[1] 31116053 31116803 31117730 31117908 31117977 31120119

Creo un vector de todos unos, con la cantidad de registros que tiene dfuturo
<br>
utilizo la instruccion  **rep()**  de R  https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/rep

In [8]:
vector_enviar <-  rep( 1,  nrow( dfuturo))

In [9]:
head( vector_enviar )

[1] 1 1 1 1 1 1

In [10]:
length( vector_enviar)

[1] 165237

Excelente, tiene tantos unos como registros tiene dfuturo

creo una tabla con  lo que voy a enviar a Kaggle

In [11]:
tabla_final  <-   as.data.table(  list(  "numero_de_cliente"= vector_ids,
                                         "Predicted"=         vector_enviar))

In [12]:
head( tabla_final)

numero_de_cliente Predicted
1 31116053          1        
2 31116803          1        
3 31117730          1        
4 31117908          1        
5 31117977          1        
6 31120119          1

In [13]:
#genero el archivo para Kaggle
#creo la carpeta donde va el experimento
dir.create( "./exp/",  showWarnings = FALSE ) 
dir.create( "./exp/ZH2016/", showWarnings = FALSE )

#genero el archivo para Kaggle
fwrite(tabla_final, 
        file= "./exp/ZH2016/todos_unos.csv",
        sep=  "," )

Subo el archivo a Kaggle y me fijo cuanta ganancia me da en el Public Leaderboard

da una ganancia de **-368.11331**   ( menos 368 millones )

### el hackeo

Ahora pasamos a calcular la cantidad de BAJA+2  que hay en los datos de septiembre-2021,  hackeamos Kaggle

Tengo dos incognitas, por un lado lo que llamo POS , los positivos, las "BAJA+2"
<br>
Por otro lado, lo que llamo NEG, los negativos,  los "BAJA+1" y "CONTINUA"

Se que se cumplen estas dos cosas:
* POS + NEG = 165237
* 117000\*POS - 3000\*NEG =-368113310

Es un sistema de dos ecuaciones con dos incógnitas, has multiples formas de resolverlo, la primera sería despejar POS.
<br>
pero utilizamos esta página  https://keisan.casio.com/exec/system/14013211335541

Al correrlo obtenemos que  POS = 1063, o sea **las BAJA+2  de 202109  son  1063**

¿Qué nos dicen estas 1063 BAJA+2 de septiembre del 2021? ¿Explican algo?
<br>
Las BAJA+2 que hay en julio del 2021 son  1268
<br>
1268 > 1063
<br>
Con lo cual es de esperar mayores ganancias de los modelos predictivos en 202107  que en 202109,  va a irnos PEOR en el Public Leaderboard que en nuestros traning/testing.